src/PHPStan/MethodAnnotationsRule.php

```diff
@@ -0,0 +1,359 @@
+<?php
+
+declare(strict_types=1);
+
+namespace SilverStripe\Standards\PHPStan;
+
+use PhpParser\Node;
+use PhpParser\Node\Stmt\Class_;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\ClassReflection;
+use PHPStan\Reflection\ReflectionProvider;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\VerbosityLevel;
+use ReflectionProperty;
+use SilverStripe\Core\Extension;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\HasManyList;
+use SilverStripe\ORM\ManyManyList;
+use SilverStripe\ORM\ManyManyThroughList;
+
+/**
+ * Validates that `@method` annotations in DataObject and Extension classes are correct,
+ * according to the relations defined within those classes.
+ *
+ * @implements Rule<Class_>
+ */
+class MethodAnnotationsRule implements Rule
+{
+    private ReflectionProvider $reflectionProvider;
+
+    public function __construct(ReflectionProvider $reflectionProvider)
+    {
+        $this->reflectionProvider = $reflectionProvider;
+    }
+
+    public function getNodeType(): string
+    {
+        return Class_::class;
+    }
+
+    public function processNode(Node $node, Scope $scope): array
+    {
+        /** @var Class_ $node */
+        $className = $node->namespacedName->toString();
+        if (!$this->reflectionProvider->hasClass($className)) {
+            return [];
+        }
+
+        $classReflection = $this->reflectionProvider->getClass($className);
+        if (!$this->isDataObjectOrExtension($classReflection)) {
+            return [];
+        }
+
+        $existingAnnotations = $this->getMethodAnnotations($classReflection);
+        $expectedAnnotations = $this->getExpectedAnnotations($classReflection);
+
+        $errors = [];
+
+        // Check that existing annotations are all valid
+        foreach ($existingAnnotations as $methodName => $annotationData) {
+            $annotationString = $annotationData['annotationString'];
+            $returnString = $annotationData['returnString'];
+
+            // Complain about any @method annotations we don't expect
+            if (!array_key_exists($methodName, $expectedAnnotations)) {
+                $errors[] = RuleErrorBuilder::message(
+                    "@method annotation '$annotationString' isn't expected and should be removed."
+                )->build();
+                continue;
```

Comment:
```suggestion
            // Complain about any @method annotations we don't expect
            $failed = false;
            if (!array_key_exists($methodName, $expectedAnnotations)) {
                $errors[] = RuleErrorBuilder::message(
                    "@method annotation '$annotationString' isn't expected and should be removed."
                )->build();
                $failed = true;
```
And then at the end of all these checks go

```php
if ($failed) {
    continue;
}
```

Means that multiple errors get captured in one go

Reply:
- If we don't expect the annotation, the only error related to it should be to remove it.
- If a method exists for an annotation, the only error related to it should be to remove it.
- If an annotation is duplicated, we won't know what other errors should be included until the duplicate is removed.